internal/wrphandlers/qos/qos.go

```diff
@@ -0,0 +1,190 @@
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt.apply(h); err != nil {
+				errs = errors.Join(errs, err)
+				h = nil
+			}
+		}
+	}
+
+	return h, errs
+}
+
+func (h *Handler) Start() error {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	if h.ctx != nil {
+		return nil
+	}
+
+	// h.queue will never block as long as the serviceQOS goroutine is running.
+	h.queue = make(chan wrp.Message, 1)
+	// h.cancel() stops serviceQOS by closing its `done` chan.
+	h.ctx, h.cancel = context.WithCancel(context.Background())
+	go serviceQOS(h.ctx.Done(), h.queue, h.maxQueueSize, h.next)
+
+	return nil
+}
+
+func (h *Handler) Stop() {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	if h.ctx == nil {
+		return
+	}
+
+	h.cancel()
+	h.ctx = nil
+	h.cancel = nil
+	// defensive: cancelling the context should be enough, but this makes things bulletproof
+	close(h.queue)
+	h.queue = nil
+}
+
+// HandleWRP queues incoming messages while the background serviceQOS goroutine attempts
+// to send as many queued messages as possible, where the highest QOS messages are prioritized
+func (h *Handler) HandleWrp(msg wrp.Message) error {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	if h.queue == nil {
+		return ErrQOSHasShutdown
+	}
+
+	// queue will never block as long as the serviceQOS goroutine is running.
+	h.queue <- msg
+
+	return nil
+}
+
+func handleWRPWrapper(next wrpkit.Handler) serviceQOSHandler {
+	return func(msg wrp.Message) (<-chan wrp.Message, <-chan struct{}) {
+		ready := make(chan struct{})
+		failedMsg := make(chan wrp.Message, 1)
+		go func() {
+			defer close(ready)
+			defer close(failedMsg)
+
+			// Note, Websocket.HandleWrp already locks between writes.
+			if err := next.HandleWrp(msg); err != nil {
+				// Delivery failed, re-enqueue message and try again later.
+				failedMsg <- msg
+				// The err itself is ignored.
+			}
+		}()
+
+		return failedMsg, ready
+	}
+}
+
+// serviceQOS is a long running goroutine that sends as many queued messages as possible,
+// where the highest QOS messages are prioritized.
+// serviceQOS starts when Handler.Start() is called.
+// serviceQOS stops when Handler.Stop() is called, closing its `done` chan.
+func serviceQOS(done <-chan struct{}, queue <-chan wrp.Message, maxQueueSize int, handleWRP serviceQOSHandler) {
+	// create and manage the priority queue
+	pq := priorityQueue{maxQueueSize: maxQueueSize}
+	var (
+		// Signaling channel from the handleWRP.
+		ready <-chan struct{}
+		// Channel for failed deliveries, re-enqueue message.
+		failedMsg <-chan wrp.Message
+	)
+
+	for {
+		select {
+		case <-done:
+			return
+		case msg, ok := <-queue:
+			if !ok {
+				// Don't enqueue an empty wrp.Message{}
+				// Handler.Stop() has been called, both `queue` and `done` are closed.
+				return
+			}
+
+			pq.Enqueue(msg)
+			if ready != nil {
+				// Previous handleWRP call has not finished, do nothing.
+			} else if top, ok := pq.Dequeue(); ok {
+				failedMsg, ready = handleWRP(top)
+			}
+		case <-ready:
```

Comment:
This is unnecessarily complex.  Just select on the queue, call `next.HandleWrp()` block the goroutine and handle the results.  Then you can delete this entire block of code.

Reply:
I think I understand the suggestion, since qos's `HandleWrp()` is already being called in this goroutine 
https://github.com/xmidt-org/xmidt-agent/blob/78dffe0cad394ab82f581940e3a8117f04941077/internal/pubsub/pubsub.go#L196-L211

then whether qos's `HandleWrp()` blocks  or not doesn't matter since it's not blocking anything

Removing this block would simplify things.

I'm cool with this idea